src/Package/Impl/ProjectSystem/RDebugLaunchProvider.cs

```diff
@@ -0,0 +1,77 @@
+using Microsoft.VisualStudio.ProjectSystem.Utilities;
+using Microsoft.VisualStudio.ProjectSystem.Utilities.DebuggerProviders;
+using Microsoft.VisualStudio.ProjectSystem.VS.Debuggers;
+
+namespace Microsoft.VisualStudio.R.Package.ProjectSystem {
+    // ExportDebugger must match rule name in ..\Rules\Debugger.xaml.
+    [ExportDebugger("RDebugger")] 
+    [AppliesTo(Constants.RtvsProjectCapability)]
+    internal class RDebugLaunchProvider : DebugLaunchProviderBase {
+        private readonly ProjectProperties _properties;
+        private readonly IRInteractiveWorkflow _interactiveWorkflow;
+
+        [ImportingConstructor]
+        public RDebugLaunchProvider(ConfiguredProject configuredProject, IRInteractiveWorkflowProvider interactiveWorkflowProvider)
+            : base(configuredProject) {
+            _properties = configuredProject.Services.ExportProvider.GetExportedValue<ProjectProperties>();
+            _interactiveWorkflow = interactiveWorkflowProvider.GetOrCreate();
+        }
+
+        private IRSession Session => _interactiveWorkflow.RSession;
+
+        public override async Task<bool> CanLaunchAsync(DebugLaunchOptions launchOptions) {
+            return Session.IsHostRunning && !string.IsNullOrEmpty(await _properties.GetStartupFileAsync());
+        }
+
+        public override Task<IReadOnlyList<IDebugLaunchSettings>> QueryDebugTargetsAsync(DebugLaunchOptions launchOptions) {
+            var targets = new List<IDebugLaunchSettings>();
+
+            if (Session.IsHostRunning) {
+                uint pid = RDebugPortSupplier.GetProcessId(Session.Id);
+
+                var target = new DebugLaunchSettings(launchOptions) {
+                    LaunchOperation = DebugLaunchOperation.AlreadyRunning,
+                    PortSupplierGuid = DebuggerGuids.PortSupplier,
+                    PortName = RDebugPortSupplier.PortName,
+                    LaunchDebugEngineGuid = DebuggerGuids.DebugEngine,
+                    ProcessId = (int)pid,
+                    Executable = RDebugPortSupplier.GetExecutableForAttach(pid),
+                };
+
+                targets.Add(target);
+            }
+
+            return Task.FromResult((IReadOnlyList<IDebugLaunchSettings>)targets);
+        }
+
+        public override async Task LaunchAsync(DebugLaunchOptions launchOptions) {
+            // Base implementation will try to launch or attach via the debugger, but there's nothing to launch
+            // in case of Ctrl+F5 - we only want to source the file. So only invoke base if we intend to debug.
+            if (!launchOptions.HasFlag(DebugLaunchOptions.NoDebug)) {
+                await base.LaunchAsync(launchOptions);
+            }
+
+            string startupFile = await _properties.GetStartupFileAsync();
+            if (!string.IsNullOrEmpty(startupFile)) {
+                _interactiveWorkflow.ActiveWindow?.Container.Show(false);
+                _interactiveWorkflow.Operations.SourceFileAsync(startupFile, echo: false)
```

Comment:
It didn't feel like there's any meaning in awaiting it - the goal of this is to kick the execution off, it shouldn't really care where it goes from there.